### PR TITLE
Seurat update

### PR DESCRIPTION
**Pull request by astrovsky01**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

Adds cite-seq functionality to seurat, permits intermediate output files, and early stops for partial runs

### CODE REVIEW

File: tools/seurat/seurat.xml [hunk 1]
```diff
@@ -1,119 +1,306 @@
...
-                <option value="PCs" selected="true">PC plots</option>
-                <option value="tsne" selected="true">tSNE plots</option>
-                <option value="heat" selected="true">Heatmap plots</option>
-            </param>
         </section>
     </inputs>
     <outputs>
         <data name="out_html" format="html" from_work_dir="out.html" label="${tool.name} on ${on_string}" />
-    </outputs>
+        <data name="norm_out" format="rdata" from_work_dir="norm_out.rds" label="${tool.name} on ${on_string}: normalization intermediate output">
```
Comment (bgruening) [inline tools/seurat/seurat.xml:176]: For what is the rdata used afterward? Would it make sense to output them conditionally?

**Comment by pavanvidem** [comment]
The current version is rigid and less practical. We probably have to make it modular like the Scanpy toolkit or even a similar set of tools as Scanpy. Then adding CITE-Seq or other modalities will be easier I guess. If you need CITE-Seq urgently, we can continue with the current structure in this update.

**Comment by astrovsky01** [comment]
It has been requested that we add it asap. I don't mind future updates relatively quickly, but if we could get this feature in sooner, that would be appreciated

File: tools/seurat/seurat.xml [hunk 2]
```diff
@@ -1,119 +1,306 @@
...
-        low_thresholds = \"${adv.low_thresholds}\",
-        high_thresholds = \"${adv.high_thresholds}\",
-        numPCs = \"${adv.num_PCs}\",
-        cells_use = \"${adv.cells_use}\",
-        resolution = \"${adv.resolution}\",
-        perplexity = \"${adv_perplexity}\",
-        min_pct = \"${adv.min_pct}\",
-        logfc_threshold = \"${adv.logfc_threshold}\",
+    #if $function.function_select == "base": 
+        Rscript -e "library(\"rmarkdown\"); render(\"$__tool_directory__/Seurat.R\",
```
[the remark this answers is not in the document]
Comment (astrovsky01) [reply tools/seurat/seurat.xml:9]: The Rscript is kind of built like that now, but it's fragile in such a way as successive commands require parameters set earlier that are never called again and are stored in the rdata object, so this was easier in my mind

File: tools/seurat/macros.xml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    </xml>
+    <xml name="cite-seq" label="Cite-seq analysis" expanded="true">
+        <param name="cite_markers" type="boolean" truevalue="T" falsevalue="F" label="Output list of cite-seq markers"/>
+        <conditional name="comparison">
+                <param name="comparison" type="select" label="Compare specific feature's effect on protein and rna expression?">
+                    <option value="yes">Yes</option>
+                    <option value="no" selected="true">No</option>
+                </param>
+                <when value="yes">
+                    <param name="feat_comp" type="text" label="Feature(s) to inspect" help="Comma-separated list of features to directly compare with protein and RNA expression"/>
```
[the remark this answers is not in the document]
Comment (astrovsky01) [reply tools/seurat/macros.xml:52]: Either a run of the main seurat tool looking at "markers" or a previous run of this cite-seq tool

File: tools/seurat/seurat.xml [hunk 4]
```diff
@@ -151,6 +361,45 @@ Rscript -e "library(\"rmarkdown\"); render(\"$__tool_directory__/Seurat.R\",
...
+                <conditional name="marker_compare">
+                    <param name="marker_compare" value="yes"/>
+                    <param name="top_x" value="3"/>
+                </conditional>
+            </conditional>
+            <section name="meta">
+                <param name="showcode" value="T"/>
+                <param name="warn" value="F"/>
+                <param name="varstate" value="F"/>
+            </section>
```
[the remark this answers is not in the document]
Comment (astrovsky01) [reply tools/seurat/seurat.xml:401]: Working on the tests right now

File: tools/seurat/seurat.xml (hunk 1 above)
Comment (astrovsky01) [reply tools/seurat/seurat.xml:176]: It was specifically requested that we output intermediate rdata to use outside of galaxy as necessary, but they are entirely optional, thus conditional

File: tools/seurat/macros.xml [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<macros>
+    <token name="@TOOL_VERSION@">4.3.0</token>
```
Comment (nsoranzo) [inline tools/seurat/macros.xml:2]:
```suggestion
    <token name="@TOOL_VERSION@">4.3.0.1</token>
```

File: tools/seurat/macros.xml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    <token name="@FUNCTION_BASE@">function['function_select'] == "base"</token>
+    <token name="@FUNCTION_CITE@">function['function_select'] == "cite"</token>
+    <token name="@VARIABLE_CONTINUE@">function['variable_continue']['variable_continue'] == "yes"</token>
+    <token name="@PCA_CONTINUE@">function['variable_continue']['pca_continue']['pca_continue'] == "yes"</token>
+    <token name="@CLUSTERS_CONTINUE@">function['variable_continue']['pca_continue']['clusters_continue']['clusters_continue'] == "yes"</token>
+    <token name="@MARKERS_CONTINUE@">function['variable_continue']['pca_continue']['clusters_continue']['markers_continue']['markers_continue'] == "yes"</token>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="@TOOL_VERSION@">r-seurat</requirement>
+            <requirement type="package" version="2.14">r-rmarkdown</requirement>
```
Comment (nsoranzo) [inline tools/seurat/macros.xml:12]:
```suggestion
            <requirement type="package" version="2.22">r-rmarkdown</requirement>
```

**Review by nsoranzo — COMMENTED** [review]
See https://github.com/galaxyproject/tools-iuc/pull/5373

**Comment by astrovsky01** [comment]
Thanks @nsoranzo!

**Review by bgruening — APPROVED** [review]

**Pull request merged**